HasOne `create_association` should raise if association already exists

zzak opened the pull request:
The goal is to raise an exception when calling `create_association` which would replace an already persisted association.

Here is a failing test report as well:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "active_support/testing/assertions"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :authors, force: true

  create_table :books, force: true do |t|
    t.integer :author_id
  end
end

class Author < ActiveRecord::Base
  has_one :book
end

class Book < ActiveRecord::Base
  belongs_to :author
  validates :author_id, uniqueness: true
end

class BugTest < Minitest::Test
  include ActiveSupport::Testing::Assertions

  def test_association_stuff
    author = Author.create!
    author.create_book!

    assert_no_difference -> { Book.count } do
      assert_raises ActiveRecord::RecordNotSaved do
        author.create_book!
      end
    end
  end
end
```

zzak committed:
Add failing test for create has_one association without dependent option

zzak committed:
Raise if dependent option is missing and target is persisted

zzak committed:
Only raise if method is nil

casperisfine commented:
> when calling create_association and not using the dependent option

I think the issue is beyond that.

This behavior makes sense if you do `author.book = Book.new(...)`, here you are explicitly replacing a `has_none` association, so it makes sense to apply the `dependent` option.

But I don't think most users expect `author.create_book` to be a shorthand for `author.book = Book.new(...)`.

I mentioned this with @fxn, and we both believe `author.create_book` should fail if the association is already set regardless of the `dependent` option.

zzak committed:
Raise if SingularAssociation#build tries to replace an existing association

zzak committed:
Only raise if calling create_record, not build

This is closer to what we want, I think, but still some test failures.. hopefully they are intentional.

zzak committed:
Need to set_new_record on BelognsToAssociation as well

zzak committed:
No need to change this test since it preserves the old behavior using build proxy

zzak committed:
Fix these tests to avoid overwriting an existing association

That is not the behavior they are testing anyways

zzak committed:
Fix has_one tests to avoid raising unnecessarily

zzak committed:
Remove unnecessary tests

zzak committed:
Fix these tests to avoid overwriting an existing association

zzak commented:
Only one failure left, it wasn't obvious to me how it should work. Would love to get some feedback on this approach, and leaving the commits for now in case that is interesting or helpful. :pray:

casperisfine commented on activerecord/lib/active_record/associations/has_one_association.rb:
```diff
@@ -84,6 +84,21 @@ def replace(record, save = true)
+
+          if target.present? && target.persisted? && owner.persisted?
+            raise RecordNotSaved.new(
+              "Failed to save the new associated #{reflection.name}.",
```
We can probably mention a record is already present.

casperisfine commented on activerecord/lib/active_record/associations/has_one_association.rb:
```diff
@@ -84,6 +84,21 @@ def replace(record, save = true)
+
+          return target unless load_target || record
+
+          if target.present? && target.persisted? && owner.persisted?
```
I don't think the `.present?` is necessary here.

casperisfine reviewed (commented):
I haven't dug very deeply, but overall it looks good to me.

My only worry here is the rollout plan, I think this behavior is the "least surprising" one, and that we should make that change.

But there is a relatively big backward compatibility concern here, so we may need to go through a deprecation cycle.

I'll defer on @matthewd on that front though, as he's much more reasonable than me on these issues 😛

lazaronixon commented:
The documentation says: https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_one

```
Account#create_beneficiary (similar to b = Beneficiary.new(account_id: id); b.save;
```

I think it was the intention in the past:

https://github.com/rails/rails/blob/b5d759fd2848146f7ee7a4c1b1a4be39e2f1a2bc/activerecord/lib/active_record/associations/has_one_association.rb#L9-L11

Are we going to have these problems again?

https://thoughtbot.com/blog/rails-has-one-limitations

All of it was fixed on 7.0.5...

<img width="714" alt="image" src="https://github.com/rails/rails/assets/2651240/66ab597e-06e2-46de-ae18-9f7cd501d78d">

casperisfine commented:
> The documentation says:

Yes, but I suspect that's not what most users expect as evidenced by https://github.com/rails/rails/issues/48330. Some users where relying on a uniqueness validator in the related model to prevent this overriding.

There is definitely a backward compatibility concern here, so we'll have to decide whether we want to go through a deprecation cycle, or just introduce that new behavior as another method.

Personally, if my `Account` `has_one :beneficiary`, I don't expect `account.create_beneficiary` to replace the existing one. I understand it as: `create_if_not_exist`.

lazaronixon commented on activerecord/lib/active_record/associations/singular_association.rb:
```diff
@@ -55,7 +55,8 @@ def set_new_record(record)
 
         def _create_record(attributes, raise_error = false, &block)
           reflection.klass.transaction do
```
This transaction is not necessary anymore.

lazaronixon commented:
> Yes, but I suspect that's not what most users expect as evidenced by https://github.com/rails/rails/issues/48330. Some users where relying on a uniqueness validator in the related model to prevent this overriding.

I think we'll have complaints about both behaviors. People who don't use unique-index or unique-validation will expect the record to be replaced. 🤷‍♂️

casperisfine commented:
> People who don't use unique-index or unique-validation will expect the record to be replaced.

I'm not sure to follow. Could you elaborate more on the logic? What does uniqueness (index or validation) changes to the problem here?

If you already "have one", why would it be logical for `create_stuff!` to just delete the data you already have. To me `create` implies there isn't already one.

I'd get the need for something like `replace_stuff!` though, seems niche but why not.

lazaronixon commented:
> People who don't use unique-index or unique-validation will expect the record to be replaced.

This has always been the Rails default behavior, "keep the integrity by deleting the old record", before and now, this is what you always had if you created a has_one without a (unique-index/unique-validation). People expecting this behavior will complain.

I think people started to think the default behavior was "prevent insertion" in reason of the proliferation of the "good practice" of "has_one should always have a unique index together with a uniqueness_validation".

What happened is that it was only possible because of a "bug" in the order of actions when creating a has_one, create_new->delete_old where the database/active_record would raise a unique constraint exception.

After fixing, delete_old->create_new, both cases "with index" or "without index" now have the default rails behavior "replace the record".

sinsoku commented:
Also check #48643 for issues with `create_association`.
Since changing the default behavior of `has_one` has a large impact, I have implemented it so that people using unique constraints or uniqueness validation can add the option.

As long as there are those who expect replacement and those who do not, it is difficult to satisfy both.
